tensorflow/core/kernels/dml_matmul_op.cc

```diff
@@ -79,57 +79,86 @@ class BaseBatchMatMulInitHelper : public InitializationHelper {
     bool adj_y;
   };
 
-  BaseBatchMatMulInitHelper(OpKernelContext* ctx,
-                            std::shared_ptr<const Attributes> attr)
+  BaseBatchMatMulInitHelper(
+      OpKernelContext* ctx, std::shared_ptr<const Attributes> attr,
+      std::function<void(OpKernelContext*, const TensorShape&,
+                         const TensorShape&)>
+          ValidateInputTensors)
       : attr_(attr) {
-    const Tensor& in0 = ctx->input(0);
-    const Tensor& in1 = ctx->input(1);
+    const TensorShape& in0_shape = ctx->input(0).shape();
+    const TensorShape& in1_shape = ctx->input(1).shape();
+    ValidateInputTensors(ctx, in0_shape, in1_shape);
+
+    TensorShape in0_batches_shape;
+    for (int i = 0; i < in0_shape.dims() - 2; ++i) {
+      in0_batches_shape.AddDim(in0_shape.dim_size(i));
+    }
+
+    TensorShape in1_batches_shape;
+    for (int i = 0; i < in1_shape.dims() - 2; ++i) {
+      in1_batches_shape.AddDim(in1_shape.dim_size(i));
+    }
+
+    BCast batches_bcast(BCast::FromShape(in0_batches_shape),
+                        BCast::FromShape(in1_batches_shape));
+
+    OP_REQUIRES(ctx, batches_bcast.IsValid(),
+                errors::InvalidArgument(
+                    "In[0] and In[1] must have compatible batch dimensions: ",
+                    in0_shape.DebugString(), " vs. ", in1_shape.DebugString()));
+
+    auto d0 = in0_shape.dim_size(in0_shape.dims() - 2);
```

Comment:
[minor] For naming, how about `in0w`, `in0h`, `in1w`, `in1h`? Seems more intuitive to me given they represent the matrices' width and height.